### ISSUE TITLE
bug: all gateways should be attempted

### ISSUE THREAD

**Issue by SgtPooki** [ipfs/ipld-explorer-components#379]
**Describe the bug**

The code is currently set to try the [gateways in serial](https://github.com/ipfs/ipld-explorer-components/blob/ca19310defb84611f55c46b1efd399217ba4a8c6/src/lib/get-raw-block.ts#L76) which means if ipfs.io takes more than timeout ms, dweb.link is never attempted. If ipfs.io takes less than timeout ms, then dweb.link is never attempted. These could be made parallel since we removed the non-delegated Helia fetch attempts.

See https://github.com/ipld/explore.ipld.io/pull/119#issuecomment-1619305060

### Proposed fix

I think for now (with only two gateways, and one method of kuboRpc fetch via kubo-rpc-client) we can run everything in parallel without issue. However, I think instead of spamming both dweb.link and ipfs.io at the same time, we should grant at least a 1/3 of timeout-ms stagger to gateway requests. 

#### Flow of `getRawBlock`:

1. Kick off all queries: kuboRpcClient & getBlockFromAnyGateway
2. Cancel all active queries if any returns a success

#### Flow of `getBlockFromAnyGateway`:

1. Query first gateway, delay request to second gateway by 1/3 timeout (instead of waiting for first query to timeout or fail)
2. Cancel active gateway queries if any return a success

**Comment by SgtPooki** [comment]
ping @whizzzkid and @lidel for agreeing on path forward.

**Comment by lidel** [comment]
So.. on explore.ipld.io, where timeout is 30s(?), 10s delay as a quick "happy eyeballs"  hack sounds okay to me.
This adds value on top of serial approach as the user may get response 0s-20s faster  if the one we got with the first request had an unlucky routing situation, or is under traffic spike.


I think we could  go one step further, and create a library that builds gateway reputation over time (sliding window) and changes the order of gateways, prioritizing the one that responds the fastest. But this could be a separate PR. 

ps. this extra heuristic is not specific to IPLD Explorer, the ipfs-geoip used on Peers screen in webui would benefit from the same thing – [currently it tries gateways in serial](https://github.com/ipfs-shipyard/ipfs-geoip/blob/466cd9d6454098c0fcf998b2217225099a654695/src/lookup.js#L18). 
Perhaps worth extracting to standalone project and reusing in both places?

**Comment by SgtPooki** [comment]
> This heuristic has a hypothesis or just intuition?

just simple brute-force simplistic approach where we have 3 methods, and we should give N methods at least 1/N time to execute

> This sounds like a straight forward Promis.any implementation, but if we're staggering calls, is the idea to wait in the promise before fetch?

pretty much a straightforward promise.any with some mapping to apply a simplistic linear delay before execution.

> I wonder what will a probabilistic fetch model look like. i.e
> * the first fetch could be kicked off together, recording winner.
> * Stagger the winner to get +N% lead with every subsequent request.
> * If the loser improves performance, the -N% lead.

I like the idea of a more intelligent approach. There are some decent operationally-light weighted sort algorithms we could implement pretty easily.

> This adds value on top of serial approach as the user may get response 0s-20s faster if the one we got with the first request had an unlucky routing situation, or is under traffic spike.

100%

> I think we could go one step further, and create a library that builds gateway reputation over time (sliding window) and changes the order of gateways, prioritizing the one that responds the fastest. But this could be a follow-up issue/PR.


created https://github.com/ipfs/ipfs-gui/issues/138

**Comment by SgtPooki** [comment]
This is fixed with the use of helia blockBrokers
